docs/framework/guides/tutorials/data-from-external-source.md

```diff
@@ -0,0 +1,877 @@
+		super( editor );
+		// property that keep the interval id
+		this.intervalId = this._intervalFetch();
+		// last fetched value
+		this.externalDataValue = '';
+	}
+
+	static get requires() {
+		return [ Widget ];
+	}
+
+	// this method will help us to clear the interval
+	destroy() {
+		clearInterval( this.intervalId );
+	}
+
+	init() {
+		this._defineSchema();
+		this._defineConverters();
+		// initial execute function to fetch and update the data
+		this._updateWidgetData();
+
+		this.editor.commands.add( 'external', new ExternalWidgetCommand( this.editor ) );
+	}
+
+	// interval function
+	_intervalFetch() {
+		return setInterval( () => this._updateWidgetData(), 15000 ); // set time interval to 15s
+	}
+
+	// fetch data and update all widget instances
+	async _updateWidgetData( externalUrl = RESOURCE_URL ) {
+		try {
+			const response = await fetch( externalUrl );
+			const data = await response.json();
+			const updateTime = new Date( data.closeTime );
+
+			// example parsed data: $17098.35 - 09/11/2022, 18:04:18
+			const parsedData = '$' + Number( data.lastPrice ).toFixed( 2 ) + ' - ' + updateTime.toLocaleString();
+
+			// update property with last fetched and parsed data
+			this.externalDataValue = parsedData;
+
+			const rootElement = this.editor.model.document.getRoot();
+
+			// loop over whole editor content, search for external widget instances
+			// and trigger `recovertItem` function
+			for ( const { item } of this.editor.model.createRangeIn( rootElement ) ) {
+				if ( item.is( 'element', 'externalElement' ) ) {
+					this.editor.editing.reconvertItem( item );
+				}
+			}
+		} catch ( error ) {
+			console.error( error );
+		}
+	}
+
+	_defineSchema() {
+		// ...
+	}
+
+	_defineConverters() {
+		// ...
+
+		editor.conversion.for( 'editingDowncast' ).elementToElement( {
+			model: 'externalElement',
+			view: ( modelElement, { writer } ) => {
+				const externalValueToShow = this.externalDataValue;
+
+				const externalDataPreviewElement = writer.createRawElement( 'span', null, function( domElement ) {
+					// when the value is not present (initial run) show a placeholder
+					domElement.textContent = externalValueToShow || 'Fetching data...';
+
+					// if new value arrive add a css animation effect to show that data were updated
+					if ( externalValueToShow ) {
+						domElement.classList.add( 'external-widget-bounce' );
+						// remove animation class when it ends
+						setTimeout( () => domElement.classList.remove( 'external-widget-bounce' ), 1100 );
+					}
+				} );
+
+				const externalWidgetContainer = writer.createContainerElement( 'span', null, externalDataPreviewElement );
+
+				return toWidget( externalWidgetContainer, writer, {
+					label: 'External widget'
+				} );
+			}
+		} );
+	}
+}
+
+```
+
+Traversing over the whole editor content can be tricky. When there relatively little content this method is sufficient, but if there is a lot of content, we suggest to use `WeakMap`.
```

Comment:
```suggestion
Editor content traversal can be a challenging process. The presented method is sufficient when there is relatively little content. Otherwise, a `WeakMap` will be a better option.
```